spring/src/test/java/io/cloudevents/spring/webflux/WebFluxRestControllerTests.java

```diff
@@ -0,0 +1,201 @@
+								+ "\"type\":\"io.spring.event\"," //
+								+ "\"source\":\"https://spring.io/events\"," //
+								+ "\"data\":{\"value\":\"Dave\"}}"),
+						String.class);
+
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+		assertThat(response.getBody()).isEqualTo("{\"value\":\"Dave\"}");
+
+		HttpHeaders headers = response.getHeaders();
+
+		assertThat(headers).containsKey("ce-id");
+		assertThat(headers).containsKey("ce-source");
+		assertThat(headers).containsKey("ce-type");
+
+		// assertThat(headers.getFirst("ce-id")).isNotEqualTo("12345");
+		assertThat(headers.getFirst("ce-type")).isEqualTo("io.spring.event.Foo");
+		assertThat(headers.getFirst("ce-source")).isEqualTo("https://spring.io/foos");
+
+	}
+
+	@Test
+	void requestResponseEvents() {
+
+		ResponseEntity<String> response = rest
+				.exchange(RequestEntity.post(URI.create("http://localhost:" + port + "/event")) //
+						.header("ce-id", "12345") //
+						.header("ce-specversion", "1.0") //
+						.header("ce-type", "io.spring.event") //
+						.header("ce-source", "https://spring.io/events") //
+						.contentType(MediaType.APPLICATION_JSON) //
+						.body("{\"value\":\"Dave\"}"), String.class);
+
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+		assertThat(response.getBody()).isEqualTo("{\"value\":\"Dave\"}");
+
+		HttpHeaders headers = response.getHeaders();
+
+		assertThat(headers).containsKey("ce-id");
+		assertThat(headers).containsKey("ce-source");
+		assertThat(headers).containsKey("ce-type");
+
+		// assertThat(headers.getFirst("ce-id")).isNotEqualTo("12345");
+		assertThat(headers.getFirst("ce-type")).isEqualTo("io.spring.event.Foo");
+		assertThat(headers.getFirst("ce-source")).isEqualTo("https://spring.io/foos");
+
+	}
+
+	@SpringBootApplication
+	@RestController
+	static class TestApplication {
+
+		@PostMapping("/")
+		public ResponseEntity<Foo> echo(@RequestBody Foo foo, @RequestHeader HttpHeaders headers) {
+			CloudEvent attributes = CloudEventHttpUtils.fromHttp(headers).withId(UUID.randomUUID().toString())
+					.withSource(URI.create("https://spring.io/foos")).withType("io.spring.event.Foo").build();
+			HttpHeaders outgoing = CloudEventHttpUtils.toHttp(attributes);
+			return ResponseEntity.ok().headers(outgoing).body(foo);
+		}
+
+		@PostMapping("/event")
+		public Mono<CloudEvent> event(@RequestBody Mono<CloudEvent> body) {
+			return body.map(event -> CloudEventBuilder.from(event).withId(UUID.randomUUID().toString())
+					.withSource(URI.create("https://spring.io/foos")).withType("io.spring.event.Foo")
+					.withData(event.getData().toBytes()).build());
+		}
+
+		@Configuration
+		public static class CloudEventHandlerConfiguration implements CodecCustomizer {
+
+			@Override
+			public void customize(CodecConfigurer configurer) {
+				configurer.customCodecs().register(new CloudEventHttpMessageReader());
+				configurer.customCodecs().register(new CloudEventHttpMessageWriter());
+			}
+
+		}
+
+	}
+
+}
+
+class Foo {
```

Comment:
Since you have this class in the other test package, maybe let's just have one copy of it in some test package, like `io.cloudevents.spring.test` in test classpath?

Reply:
I'm not sure. It's just a demo POJO. There's no reason to share it IMO.

Reply:
It's just do deduplicate that code

Reply:
There's no duplication. The two Foos are not related in any way.